### PR TITLE
(Truffle optimizations) Final truffle optimizations

### PR DESCRIPTION
**Pull request by Alexzerntev**
- Separated pure records from records with duplicate keys
- Made more PE friendly nullables and tryables
- Fixed record operations to have a specialization instance per field per operation (as truffle docs suggest)
- Added toString() overrides for better debugging
- Merged TreeMap implementation to get and put with one walk
- Added missing TruffleBoundaries
- Removed language and context storage from runtime objects. It can be accessed from inside the nodes (also made it cached)

### CODE REVIEW

File: snapi-truffle/src/main/java/raw/runtime/truffle/runtime/record/DuplicateKeyRecord.java [hunk 1]
```diff
@@ -12,46 +12,79 @@
...
-  private boolean validDistinctKeys = true;
-
-  public RecordObject(Shape shape) {
+  public DuplicateKeyRecord(Shape shape) {
     super(shape);
-    this.distinctKeys = new Vector<>();
+    updateDistinctKeys();
+  }
+
+  private void updateDistinctKeys() {
```
Comment (Alexzerntev) [inline snapi-truffle/src/main/java/raw/runtime/truffle/runtime/record/DuplicateKeyRecord.java:43]: Add @TruffleBoundary

**Comment by Alexzerntev** [comment]
Merging into initial PR (https://github.com/raw-labs/snapi/pull/352) . The review will be done here.

File: snapi-truffle/src/main/java/raw/runtime/truffle/PropertyType.java [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0, included in the file
+ * licenses/APL.txt.
+ */
+
+package raw.runtime.truffle;
+
+// Static properties for dynamic object storage types
+public class PropertyType {
+  public static final byte INT_TYPE = 0;
```
Comment (Alexzerntev) [inline snapi-truffle/src/main/java/raw/runtime/truffle/PropertyType.java:17]: To check for short and byte

File: snapi-truffle/src/main/java/raw/runtime/truffle/ExpressionNode.java [hunk 3]
```diff
@@ -97,8 +98,13 @@ public LocationObject executeLocation(VirtualFrame virtualFrame)
     return RawTypesGen.expectLocationObject(executeGeneric(virtualFrame));
   }
 
-  public RecordObject executeRecord(VirtualFrame virtualFrame) throws UnexpectedResultException {
-    return RawTypesGen.expectRecordObject(executeGeneric(virtualFrame));
+  public PureRecord executePureRecord(VirtualFrame virtualFrame) throws UnexpectedResultException {
+    return RawTypesGen.expectPureRecord(executeGeneric(virtualFrame));
+  }
+
+  public DuplicateKeyRecord executeConcatRecord(VirtualFrame virtualFrame)
```
Comment (Alexzerntev) [inline snapi-truffle/src/main/java/raw/runtime/truffle/ExpressionNode.java:105]: Typo

**Review by bgaidioz — APPROVED** [review]
We did review it together.

**Pull request merged**